crates/networking/manager/src/gossipsub/validate/beacon_block.rs

```diff
@@ -0,0 +1,291 @@
+            [signed_bls_execution_change.message.validator_index as usize];
+
+        cached_db.seen_bls_to_execution_signature.write().await.put(
+            AddressSlotIdentifier {
+                address: validator.public_key.clone(),
+                slot: block.message.slot,
+            },
+            signed_bls_execution_change.message.clone(),
+        );
+    }
+
+    Ok(ValidationResult::Accept)
+}
+
+pub async fn validate_beacon_block(
+    beacon_chain: &BeaconChain,
+    cached_db: &CachedDB,
+    block: &SignedBeaconBlock,
+    state: &BeaconState,
+    is_parent: bool,
+) -> anyhow::Result<ValidationResult> {
+    let store = beacon_chain.store.lock().await;
+
+    // [IGNORE] The block is not from a future slot.
+    if block.message.slot > store.get_current_slot()? {
+        return Ok(ValidationResult::Ignore(
+            "Block is from a future slot".to_string(),
+        ));
+    }
+
+    // [IGNORE] The block is from a slot greater than the latest finalized slot.
+    if block.message.slot
+        <= compute_start_slot_at_epoch(store.db.finalized_checkpoint_provider().get()?.epoch)
+    {
+        return Ok(ValidationResult::Ignore(
+            "Block is from a slot greater than the latest finalized slot".to_string(),
+        ));
+    }
+
+    let validator =
+        if let Some(validator) = state.validators.get(block.message.proposer_index as usize) {
+            validator
+        } else {
+            return Ok(ValidationResult::Reject("Validator not found".to_string()));
+        };
+
+    // [IGNORE] The block is the first block with valid signature received for the proposer for the
+    // slot.
+    if cached_db
+        .seen_proposer_signature
+        .read()
+        .await
+        .contains(&AddressSlotIdentifier {
+            address: validator.public_key.clone(),
+            slot: block.message.slot,
+        })
+    {
+        return Ok(ValidationResult::Ignore(
+            "Signature already received".to_string(),
+        ));
+    }
+
+    // [REJECT] The proposer signature, signed_beacon_block.signature, is valid with respect to the
+    // proposer_index pubkey.
+    match state.verify_block_header_signature(&block.signed_header()) {
+        Ok(true) => {}
+        Ok(false) => {
+            return Ok(ValidationResult::Reject("Invalid signature".to_string()));
+        }
+        Err(err) => {
+            return Ok(ValidationResult::Reject(format!(
+                "Signature verification failed: {err}"
+            )));
+        }
+    }
+
+    match store
+        .db
+        .beacon_block_provider()
+        .get(block.message.parent_root)?
+    {
+        Some(parent_block) => {
+            // [REJECT] The block is from a higher slot than its parent.
+            if block.message.slot > parent_block.message.slot + 1 {
+                return Ok(ValidationResult::Reject(
+                    "Block is from a higher slot than expected".to_string(),
+                ));
```

Comment:
```suggestion
    match store
        .db
        .beacon_block_provider()
        .get(block.message.parent_root)?
    {
        Some(parent_block) => {
            // [REJECT] The block is from a higher slot than its parent.
            if block.message.slot <= parent_block.message.slot {
                return Ok(ValidationResult::Reject(
                    "Block is from a higher slot than expected".to_string(),
                ));
```

^ shouldn't the condition be this?